docs/rfc/rfc-010-p2p-light-client.rst

```diff
@@ -0,0 +1,127 @@
+LibP2P project, it's useful to think about the ways that these efforts may
+combine or interact.
+
+We expect to be able to leverage libp2p tools to provide some kind of service
+discovery for tendermint-based networks. This means that it will be possible
+for the p2p stack to easily identify specialized nodes, (e.g. light clients)
+thus obviating many of the design challenges with providing this feature in
+the context of the current stack.
+
+Similarly, libp2p makes it possible for project to be able back their non-Go
+light clients, without the major task of first implementing tendermint's p2p
+connection handling. We should identify if there exist users (e.g. the go IBC
+relayer, it's maintainers, and operators) who would be able to take advantage
+of p2p light client, before switching to libp2p. To our knowledge there are no
+external implementations of this p2p protocol, and it seems unlikely that a
+team would implement this directly ahead of its impending removal.
+
+User Cases
+~~~~~~~~~~
+
+This RFC makes a few assumptions about the use cases and users of light
+clients in tendermint.
+
+One of the most active use cases for light clients is in the implementation of
+the IBC relayer. Thus, we expect that providing P2P light clients might
+increase the reliability of relayers and reduce the cost of running a relayer,
+because relayer operators won't have to decide between rely on public RPC
+endpoints (unreliable) or running their own full nodes (expensive.) This
+also assumes that there are *no* other uses of the RPC in the relayer, and
+unless the relayers have the option of dropping all RPC use, it's unclear if a
+P2P light client will actually be able to successfully remove the dependency
+on the RPC system.
+
+Given that the primary relayer implementation, is hermes (rust,) it might be
+safe to deliver a version of tendermint that adds a light client rector in
+the full nodes, but that does not provide an implementation of a Go light
+client. This either means that the rust implementation would need support for
+the legacy P2P connection protocol or wait for the libp2p implementation.
+
+Client side light client (e.g. wallets, etc.) users may always want to use (a
+subset) of the RPC rather than connect to the P2P network for an ephemeral
+use.
+
+Discussion
+----------
+
+Implementation Questions
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+Most of the complication in the is how to have a long lived light client node
+that *only* runs the light client reactor, as this raises a few questions:
+
+- would users specify a single P2P node to connect to when creating a light
+  client or would they also need/want to discover peers?
+
+- how do we prevent full nodes and validators from allowing their peer slots,
+  which are typically limited, from filling with light clients? If
+  light-clients aren't limited, how do we prevent light clients from consuming
+  resources on consensus nodes?
```

Comment:
This is a good question. Also we need to consider that the usage of light clients varies. In the case of a relayer, we can expect the light client to pretty much consistently be requesting headers. In the case of a in-browser wallet we'd expect infrequent flurries of requests. For the relayer case do we just treat them as a full-node? For the in-browser wallet case do we open and close the connection only when the browser is being used?

Reply:
I would guess that the in-browser use would want to be RPC based, though I don't know.

I think further defining these use cases will help us build the right thing.

Reply:
> I would guess that the in-browser use would want to be RPC based, though I don't know.

Most likely - or at least they will probably need to use RPC to fetch the other forms of data. Although, I think the SDK team are trying to push using gRPC for all state queries so there could be a path where browsers don't use Tendermint's RPC at all

Reply:
doing grpc from the browser isn't such a walk in the park either. In any case this would be grpc/rpc from the application rather than from tendermint, correct? If so while we might want to prevent light clients from needing to use the tendermint RPC, it seems like specific chains (for the browser case) would still want a light client that talked to *that* RPC which seems like a possibility.

Reply:
Yeah I think the assumption that browsers will still want to use the light client RPC server is valid. I'm not sure if that means that they would also want verification to be done via RPC or if there would be any advantages to doing it via the P2P layer.

Another angle is that getting information only needs to be done with one peer. i.e. a single RPC address whereas verifying that data requires being connected to multiple RPC addresses. Using the P2P layer means you don't need to worry about finding multiple addresses you just need the one. (I'm not sure how much of a benefit that is though).

I've often asked myself that if networks had an abundance of public RPC endpoints whether there would be that much of a need to support P2P.

I guess my current inclination is that having both available is a good idea.

Reply:
Having both is, I think good, and it seems reasonable to do some product research on this question *after* libp2p, and after the SDK's gRPC service(s) mature a bit.

Reply:
I've added a potential answer to the text here, based on this conversation.